packages/mds-event-processor/src/proc-event.ts

```diff
@@ -0,0 +1,269 @@
+      const deviceState = {
+        ...baseDeviceState,
+        annotation: annotation,
+        gps: gps,
+        service_area_id: service_area_id,
+        charge: charge,
+        state: EVENT_STATUS_MAP[event_type as VEHICLE_EVENT],
+        event_type: event_type,
+        event_type_reason: event_type_reason,
+        trip_id: trip_id
+      } as StateEntry
+      // Take necessary steps on event trasitions
+      switch (data.event_type) {
+        case 'trip_start':
+          processTripEvent(deviceState)
+        case 'trip_enter':
+          processTripEvent(deviceState)
+        case 'trip_leave':
+          processTripEvent(deviceState)
+        case 'trip_end':
+          processTripEvent(deviceState)
+      }
+      // Only update cache (device:state) with most recent event
+      if (!lastState || lastState.timestamp < deviceState.timestamp) {
+        await cache.hset('device:state', provider_id + ':' + device_id, JSON.stringify(deviceState))
+      }
+      // Add to PG table (reports_device_states) and stream
+      await db.insert('reports_device_states', deviceState)
+      //await stream.writeCloudEvent('mds.processed.event', JSON.stringify(device_state))
+      return deviceState
+    }
+
+    case 'telemetry': {
+      const { gps, charge } = data
+      const annotation = getAnnotationData(gps)
+      const district = annotation.areas.length ? annotation.areas[0].id : null
+      const deviceState = {
+        ...baseDeviceState,
+        annotation: annotation,
+        gps: gps,
+        charge: charge
+      } as StateEntry
+      // Match telemetry to trip data
+      processTripTelemetry(deviceState)
+
+      // Add to PG table (reports_device_states) and stream
+      await db.insert('reports_device_states', deviceState)
+      //await stream.writeCloudEvent('mds.processed.event', JSON.stringify(device_state))
+      return deviceState
+    }
+  }
+}
+
+async function processTripEvent(deviceState: StateEntry) {
+  /*
+  Add vehicle events of a trip to cache (trips:events):
+
+    Key: 'provider_id:device_id'
+    Value: hash map of tripEvents keyed by trip_id
+
+  */
+  const {
+    timestamp,
+    event_type,
+    event_type_reason,
+    annotation_version,
+    annotation,
+    gps,
+    service_area_id,
+    trip_id,
+    provider_id,
+    device_id
+  } = deviceState
+
+  const tripEvent = {
+    vehicle_type: 'scooter',
+    timestamp: timestamp,
+    event_type: event_type,
+    event_type_reason: event_type_reason,
+    annotation_version: annotation_version,
+    annotation: annotation,
+    gps: gps,
+    service_area_id: service_area_id
+  } as TripEvent
+
+  // Either append to existing trip or create new entry
+  const cacheEntry: string | null = await cache.hget('trips:events', provider_id + ':' + device_id)
+  let trips: { [trip_id: string]: TripEvent[] } = cacheEntry ? JSON.parse(cacheEntry) : {}
+  //TODO reduce logic
```

Comment:
I haven't looked very deeply at the relevant types/data shapes, but I'd think we could probably do some kind of spread assignment here.